### PR TITLE
nixos/nixosSystem: Remove deprecated arguments, deprecate pkgs argument, and add errors to specialArgs

### PR DESCRIPTION
**Pull request by Gerg-L**
## Description of changes
passing `pkgs` to `specialArgs` and `lib.nixosSystem` is a common anti-pattern that new users fall into which causes unexpected behavior and complete ignorance to the `nixpkgs.` option set

Since [readOnlyPkgs](https://github.com/NixOS/nixpkgs/commit/e5db80ae487b59b4e9f950d68983ffb0575e26c6) has merged there's a easy immutable way to set `pkgs`

I think it's time to do away with alternative handling methods of `pkgs`

opinions are appreciated, thanks

### CODE REVIEW

**Comment by Gerg-L** [comment]
TODO: add documentation

File: nixos/lib/eval-config.nix [hunk 1]
```diff
@@ -57,14 +53,12 @@ let
         # they way through, but has the last priority behind everything else.
         nixpkgs.system = lib.mkDefault system;
       })
-      ++
-      (optional (pkgs_ != null) {
-        _module.args.pkgs = lib.mkForce pkgs_;
-      })
```
Comment (roberth) [inline nixos/lib/eval-config.nix:63]: We shouldn't change the behavior with just a warning, and I feel like this is too big a change anyway - as this is the stable NixOS entrypoint.

How about we start with removing `pkgs` from `lib.nixosSystem`? That's a weird one, where it's unclear what the intent should even be: ignore `nixpkgs.*`? Use `readOnlyPkgs`? I think both are equally surprising and users should make a choice (if they pass `pkgs` at all).

**Review by roberth — CHANGES_REQUESTED** [review]
- A bit too aggressive on the stable api.
- Secondarily: don't tolerate error conditions with just a warning.

Also seems like a good idea to finish https://github.com/NixOS/nixpkgs/pull/231940 before triggering deployment tool maintainers to look at their `nixpkgs.*` situation.

**Comment by Gerg-L** [comment]
I mostly agree with you roberth

Mainly put this out there to get opinions

File: nixos/lib/eval-config.nix (hunk 1 above)
Comment (Gerg-L) [reply nixos/lib/eval-config.nix:63]: My intent is making old guides which use a pattern that looks like this:

```nix
  outputs = { nixpkgs, ... }:
  let
    system = "x86_64-linux";
    pkgs = import nixpkgs {
      inherit system;
      config.allowUnfree = true;
   };
  in {
    nixosConfigurations.hostname = nixpkgs.lib.nixosSystem {
      inherit system;
      inherit pkgs; #this
      specialArgs = { inherit pkgs; };#or this
      };
    }
```
Stop shooting people in the foot

**Comment by Gerg-L** [comment]
i'll make a similar PR later

**Pull request closed without merging**